### PR TITLE
fix: ignore heredoc prose in GateGuard

### LINKED ISSUE #2886: gateguard destructive-command detector matches heredoc body text, not just the actual command
Ran into this while writing a migration doc — a plain `cat > notes.md <<'EOF' ... EOF` command got blocked because the heredoc content (documentation text, nothing being executed) happened to contain words the destructive-command matcher looks for.

Traced it to `isDestructiveBash()` in `scripts/hooks/gateguard-fact-force.js` (~line 669-715):

- `DESTRUCTIVE_SQL_DD` (`drop table|delete from|truncate|dd if=`) is regex-tested against `flattened = explodeSubshells(stripQuotedStrings(raw))`, which is basically the whole command text. `stripQuotedStrings` blanks out `'...'` and `"..."` but a heredoc body isn't inside quotes, so none of that text gets stripped before the regex runs.
- On the `rm` side, `isDestructiveRm()` is token-based (needs the segment's first token to literally be `rm`), but segment splitting (`splitCommandSegments`) only splits on `;|&` and has no concept of heredocs. If the heredoc body happens to contain a `;` followed by prose that starts with "rm -rf ...", that also gets read as its own segment and can trip the rm check too.

Repro:

```
cat > migration-notes.md <<'EOF'
This migration will DROP TABLE old_sessions once we've verified nothing reads from it anymore.
EOF
```

That's just writing a markdown file. Nothing here touches a database, but the gate treats the phrase sitting inside the heredoc the same as if it were literally the command being run.

There's already a precedent for this exact class of bug: #352 was the same root cause (heredoc body text getting matched as if it were the real invocation) but in `pre-bash-dev-server-block.js` against the dev-server pattern, and it got fixed there. The destructive-command matcher in `gateguard-fact-force.js` has the same gap and never got the equivalent fix.

Fix direction: strip heredoc bodies out of `raw` before any of the destructive regex/tokenizing runs (same idea as `stripQuotedStrings`, just for `<<'EOF' ... EOF` / `<<EOF ... EOF` / `<<-EOF ... EOF` blocks), or alternatively only match against tokens that are actually in command position instead of running `.test()` against the whole flattened text. Happy to send a PR if that direction sounds reasonable — mostly wanted to flag it since I hit it doing ordinary doc writing, nothing exotic.

File: `scripts/hooks/gateguard-fact-force.js`, function `isDestructiveBash()`. Checked against current `main` on the local plugin checkout.

### PR DESCRIPTION
**Pull request by dajiaohuang**
## What Changed

- Strip confidently parsed heredoc payloads before GateGuard classifies destructive commands.
- Preserve executable substitutions from unquoted heredocs while treating quoted heredoc payloads as inert data.
- Fail closed on ambiguous delimiters, arithmetic/quote contexts, and line continuations.
- Add regression coverage for quoted/unquoted/tab-stripping heredocs and adversarial parser cases.

## Why This Change

GateGuard currently classifies prose in heredoc bodies as shell commands, so ordinary documentation writes can trigger destructive-command denials. This fixes #2886 without allowing actual commands before/after a heredoc or substitutions inside unquoted payloads to bypass the gate.

## Testing Done

- [x] Focused GateGuard suite: 166 passed, 0 failed (`node tests/hooks/gateguard-fact-force.test.js`)
- [x] ESLint passed for both changed files
- [x] `node --check scripts/hooks/gateguard-fact-force.js`
- [x] `git diff --check`
- [x] Edge cases independently code- and security-reviewed; final reviews clean
- [ ] Full suite passes locally (`node tests/run-all.js`): 3847/3877 passed; 30 unrelated Windows-environment failures require WSL/symlink support or isolated package resolution and do not touch these files

## Type of Change

- [x] `fix:` Bug fix
- [x] `test:` Tests

## Security & Quality Checklist

- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly (not applicable; no JSON changed)
- [x] Shell scripts pass shellcheck (not applicable; JavaScript only)
- [x] Pre-commit hooks pass locally (no configured hook changed the commit)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation

- [x] Added comments for the non-trivial parsing and fail-closed behavior
- [x] README update not needed

### CODE REVIEW

File: scripts/hooks/gateguard-fact-force.js [hunk 1]
```diff
@@ -151,6 +151,168 @@ function stripQuotedStrings(input) {
...
+    }
+
+    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(delimiter)) return null;
+    const next = line[i + 1];
+    if (next && !/[\s;&|<>()]/.test(next)) return null;
+    heredocs.push({ delimiter, quoted, stripTabs });
+  }
+
+  return quote || escaped ? null : heredocs;
+}
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-fact-force.js]: Addressed in 48734196: moved the heredoc parser into scripts/hooks/gateguard-heredoc.js, split delimiter parsing from the line scanner, and added multi-heredoc plus fail-closed security regressions. Focused GateGuard tests pass 171/171; the npm publish-surface test passes 2/2.

File: scripts/hooks/gateguard-fact-force.js [hunk 2]
```diff
@@ -151,6 +151,168 @@ function stripQuotedStrings(input) {
...
+  return [...substitutions];
+}
+
+/**
+ * Remove heredoc payload text before classifying the surrounding shell
+ * command. Prose in a heredoc is data, so matching it as a command produces
+ * false positives. Unquoted heredocs can still execute `$()` and backtick
+ * substitutions; retain the complete payload whenever either syntax appears.
+ * Quoted heredoc delimiters disable expansion, so their payload is fully inert.
+ * Ambiguous shell syntax returns the original input unchanged (fail closed).
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-fact-force.js]: Addressed in 48734196: the helper comment now states that only executable substitution bodies are retained and the remaining unquoted payload is dropped. Focused GateGuard tests pass 171/171.

File: scripts/hooks/gateguard-heredoc.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      escaped = true;
+      continue;
+    }
+    if (ch === '`' || (ch === '$' && text[i + 1] === '(')) {
+      for (const substitution of extractCommandSubstitutions(text.slice(i))) {
+        substitutions.add(substitution);
+      }
+    }
+  }
+  return [...substitutions];
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js]: Addressed in d56b299b: the parser now uses generator-backed discovery and line-index spans instead of mutating Sets or arrays, while preserving ordering, deduplication, and fail-closed behavior. Validation: GateGuard 171/171, npm publish surface 2/2, ESLint, Prettier, syntax, and diff checks passed; 250,015 differential cases matched the prior behavior, and a 40k-line heredoc remained linear at about 2.6 ms.

File: scripts/hooks/gateguard-heredoc.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    }
+    if (ch === '\\') {
+      escaped = true;
+      continue;
+    }
+    if (ch === '`' || (ch === '$' && text[i + 1] === '(')) {
+      yield* extractCommandSubstitutions(text.slice(i));
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js]: Handled in d0a05a19. Heredoc payloads now call the shared extractor once with literal outer-quote semantics; nested substitutions retain normal shell quoting. I also removed repeated immutable-array copying across multiple heredocs. Validation: shell-substitution 53/53, GateGuard 171/171, and a 4,000-substitution body improved from about 1.3 s to 1.4 ms; the 48,000-heredoc worst case remains at baseline around 184 ms.

File: scripts/hooks/gateguard-heredoc.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    const heredocs = findHeredocs(line);
+    if (heredocs === null) return raw;
+    if (heredocs.length > 0 && !isProvenPassiveHeredocLine(line)) return raw;
+    if (heredocs.length > 0) {
+      pending = heredocs;
+      pendingIndex = 0;
+      bodyStartIndex = lineIndex + 1;
+      headerIndex = lineIndex;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js:247]: Handled in d0a05a19. Body consumption is now delegated to focused helpers; every changed parser function is under 50 lines, nesting stays within the repository limit, and changed code performs no direct object/array mutation. Both code and security specialist reviews are clean; focused tests pass (53/53 shell substitution, 171/171 GateGuard).

File: scripts/hooks/gateguard-heredoc.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+          trailingStartIndex = lineIndex + 1;
+        }
+      }
+      continue;
+    }
+    if (completedHeredoc && line.trim()) return raw;
+    if (completedHeredoc) continue;
+    const heredocs = findHeredocs(line);
+    if (heredocs === null) return raw;
+    if (heredocs.length > 0 && !isProvenPassiveHeredocLine(line)) return raw;
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js:241]: Verified, but I am not changing this in this PR. A quoted heredoc payload is stored data, not immediately executed shell syntax; retaining it would recreate #2886 false positives, including the tested `cat > migration-notes.md <<'EOF'` case. GateGuard currently allows ordinary redirection generally (the equivalent `printf ... >> ~/.bashrc` is also outside its destructive-command policy), so protecting startup-file targets would be a separate delayed-execution policy with narrowly scoped tests, not a heredoc-filtering fix. Independent security review reached the same conclusion.

File: scripts/hooks/gateguard-heredoc.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+  let headerIndex = -1;
+  let trailingStartIndex = lines.length;
+  let substitutionText = '';
+  let substitutionCount = 0;
+  let completedHeredoc = false;
+  for (let lineIndex = 0; lineIndex < lines.length; lineIndex += 1) {
+    const line = lines[lineIndex];
+    if (pendingIndex < pending.length) {
+      const current = pending[pendingIndex];
+      if (!current.quoted && /\\$/.test(line)) return raw;
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js]: Handled in 8deba21a. Unquoted heredoc physical lines are now joined using Bash odd-backslash continuation semantics before substitution classification; `<<-` strips only leading tabs after the join, preserving internal continuation tabs. Added red-then-green regressions for split command names, split options under `<<-`, and the corresponding internal-tab allow case. Verified against Git Bash 5.2, with GateGuard 174/174, shell-substitution 53/53, lint/format/diff checks, npm publish-surface 2/2, and independent code/security reviews all clean.

File: scripts/hooks/gateguard-heredoc.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+ *
+ * @param {string[]} lines
+ * @param {number} startIndex
+ * @param {{ delimiter: string, quoted: boolean, stripTabs: boolean }} heredoc
+ * @returns {{ nextIndex: number, substitutions: string[] } | null}
+ */
+function consumeHeredocBody(lines, startIndex, heredoc) {
+  for (let lineIndex = startIndex; lineIndex < lines.length; lineIndex += 1) {
+    const line = lines[lineIndex];
+    if (!heredoc.quoted && /\\$/.test(line)) return null;
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js]: Fixed in 8deba21a. This review ran against d0a05a19 before the continuation patch. The current head joins odd-backslash physical continuations before extracting unquoted heredoc substitutions and includes red-then-green regressions for a split `rm` name plus the `<<-` split-option variant. Git Bash 5.2 oracle parity, GateGuard 174/174, shell-substitution 53/53, and independent final code/security reviews are clean.

File: scripts/hooks/gateguard-heredoc.js [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+/** @returns {string} */
+function normalizeUnquotedHeredocLines(lines, stripTabs = false) {
+  const logical = lines
+    .map((line, index) => {
+      if (index === lines.length - 1) return line;
+      return hasLineContinuation(line) ? line.slice(0, -1) : `${line}\n`;
+    })
+    .join('');
+  return stripTabs ? logical.replace(/^\t+/, '') : logical;
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply scripts/hooks/gateguard-heredoc.js:141]: Verified against GNU Bash 5.2.37. For a `<<-` body containing tab-indented `A\` followed by tab-indented `B`, `od -An -tx1` emits `41 09 42 0a`: the second line's tab survives as an internal tab after the continuation joins the lines. The equivalent split-command oracle attempts `p` with `rintf` as a separate word, not `printf`. The current normalization and `expectAllow` case therefore match Bash behavior, so I’m retaining them.

File: tests/hooks/gateguard-fact-force.test.js [hunk 10]
```diff
@@ -1477,6 +1477,395 @@ function runTests() {
...
+  if (
+    test('preserves internal tabs after tab-stripped heredoc continuations', () => {
+      expectAllow(
+        ['cat <<-EOF', '\t$(r\\', '\tm -rf /tmp/expanded-target)', 'EOF'].join('\n'),
+        'internal tab after tab-stripped heredoc continuation'
+      );
+    })
+  )
+    passed++;
+  else failed++;
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/hooks/gateguard-fact-force.test.js:1794]: Verified the scope against the current branch and upstream `main`: the test file is already 2,238 lines upstream and 2,597 lines here, while the heredoc block spans 356 lines. Extracting only this PR’s heredoc cases would still leave roughly 2,241 lines, so it would not satisfy the requested 800-line limit. Reaching that threshold requires an unrelated refactor of most of the pre-existing suite; I’m keeping this security fix focused.
